client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -541,6 +543,10 @@ private void failConnectionAttempt(final Throwable connectionFailureCause, final
         } else if (connectionFailureCause instanceof MechanismMismatchException) {
             connectionHandler.handle(Future.failedFuture(
                     new ClientErrorException(HttpURLConnection.HTTP_UNAUTHORIZED, "no suitable SASL mechanism found for authentication with server")));
+        } else if (connectionFailureCause instanceof SSLException) {
+            connectionHandler.handle(Future.failedFuture(
+                    new ClientErrorException(HttpURLConnection.HTTP_UNAUTHORIZED,
```

Comment:
Wouldn't a `400 Bad Request` be more appropriate here?
(Possibly `401 Unauthorized` could be tied to the `SSLPeerUnverifiedException` subclass of `SSLException`, although I wonder if that can occur here.)

Reply:
I gave that a lot of thought as well :-) You could also try to use a more specific exception for `SSLKeyException`. However, since it is used to report problems with both the client as well as the server cert, you don't exactly know if it should be a `ClientErrorException` or a `ServerErrorException`.
I had also thought about just using a 400 code but again, the problem may also lie on the server side, so it is hard to do the exact right thing here.
I am fine with using the 400 code but I think we should stick with a single exception (and code) that we use for all TLS related problems for simplicity. In the end, we are also including the root cause which might be used to better understand the underlying problem. In any case, the SSLException should be considered a terminal condition ...

Reply:
Ok, I would opt for using `400 Bad Request` as the single exception then.
Otherwise an "unauthorized" exception could be confusing in the case that there is an issue with incompatible cipher suites for example.

Reply:
done